substrate/frame/support/src/traits/tokens/fungible/regular.rs

```diff
@@ -181,8 +182,9 @@ pub trait Unbalanced<AccountId>: Inspect<AccountId> {
 	) -> Result<Self::Balance, DispatchError> {
 		let old_balance = Self::balance(who);
 		let free = Self::reducible_balance(who, preservation, force);
-		if let BestEffort = precision {
-			amount = amount.min(free);
+		match precision {
+			BestEffort => amount = amount.min(free),
+			Exact => ensure!(free >= amount, TokenError::FundsUnavailable),
```

Comment:
this is better than what we have now, so good enough to merge it.
but we should find a way to differentiate between `FundsUnavailable` and causes account to be deleted